[DOCS] Update Docker Compose setup (#1906)
elastic/stack-docs: docs/en/getting-started/get-started-docker.asciidoc

. Install the appropriate https://docs.docker.com/get-docker/[Docker application]
for your operating system.

> . If you're running on Linux, install https://docs.docker.com/compose/install/[Docker Compose].

Reviewer: This should actually be in the Docker Compose Prerequisites section, not here (Docker).

Author: ++ I'll move that line to the Docker Compose prerequisites.

To get a multi-node {es} cluster and {kib} up and running in Docker with
security enabled, you can use Docker Compose.

This configuration provides a simple method of starting a secured cluster that
you can use for development before building a distributed deployment with
multiple hosts.

[discrete]
==== Prerequisites

> 

Author: I'll add another bullet to the Prerequisites:

>. If you're running on Linux, install https://docs.docker.com/compose/install/[Docker Compose].

referenced by the `docker-compose.yml` file.

["source","txt",subs="attributes"]
----
include::docker/.env[]
----

[discrete]
[[docker-compose-file]]
===== `docker-compose.yml`

The `docker-compose.yml` file is a Docker Compose file that defines generates
node security certificates for {es}, configures a three-node {es} cluster with
TLS enabled for client and internode communications, and securely enrolls a {kib}
instance with the cluster. 

.Exposing ports on Linux
****
This configuration exposes port `9200` on all network interfaces. Given how
Docker manipulates `iptables` on Linux, this means that your {es} cluster is
publicly accessible, potentially ignoring any firewall settings. If you don't
want to expose port `9200` and instead use a reverse proxy, replace `9200:9200`
with `127.0.0.1:9200:9200` in the `docker-compose.yml` file. {es} will then only
be accessible from the host machine itself.
****

[source,yaml,subs="attributes"]
----
include::docker/docker-compose.yml[]
----

endif::[]
--

[discrete]
==== Start your cluster with security enabled and configured

. Modify the `.env` file and enter a password value for the `ELASTIC_PASSWORD`

> and `KIBANA_PASSWORD` variables.

Author: I'll incorporate these changes in my next commit.

. Create and start the three-node {es} cluster:
+
["source","sh"]
----
docker-compose -f docker-compose.yml up -d
----

. Open a browser and navigate to http://localhost:5601[http://localhost:5601] to
access {kib}, where you can load sample data and interact with your cluster.

[[docker-compose-remove-containers]]
[discrete]
==== Remove Docker Compose containers

When you're done experimenting, you can remove the network, containers, and
volumes:

[source,sh]
----
docker network rm <directory-name>_default
docker-compose -f docker-compose.yml down --volumes
----

`directory-name`:: Name of the directory that you created for the `.env` and
`docker-compose.yml` files.

[discrete]
[[load-settings-file]]
==== Load settings from a file

Specifying settings for {es} and {kib} directly in the Docker Compose file is a
convenient way to get started, but loading settings from a file is preferable
after you get past the experimental stage.

> For example, to use `es-node01.yml` as the configuration file for the `es-node01`

Reviewer: ```suggestion
For example, to use a custom `es01.yml` as the configuration file for the `es01`
```

> {es} node, you can create a bind mount in the volumes section.

Reviewer: ```suggestion
{es} node, you can create a bind mount in the **volumes** section for the **es01** service.
```

the cluster. The `KIBANA_PASSWORD` value is only used internally when
configuring {kib}.

. Create and start the three-node {es} cluster and {kib} instance:
+
["source","sh"]
----
docker-compose up -d
----

. When the deployment has started, open a browser and navigate to http://localhost:5601[http://localhost:5601] to
access {kib}, where you can load sample data and interact with your cluster.

[[docker-compose-remove-containers]]
[discrete]
==== Remove Docker Compose containers

When you're done experimenting, you can remove the network, containers, and
volumes:

[source,sh]
----
docker-compose down -v
----

[discrete]
[[load-settings-file]]
==== Load settings from a file

Specifying settings for {es} and {kib} directly in the Docker Compose file is a
convenient way to get started, but loading settings from a file is preferable
after you get past the experimental stage.

For example, to use a custom `es01.yml` as the configuration file for the `es01`
{es} node, you can create a bind mount in the `volumes` section for the `es01`
service.

["source","yaml"]
----
    volumes:
      - data01:/usr/share/elasticsearch/data
      - certs:$CERTS_DIR

>       - ./es01.yml:/usr/share/elasticsearch/config/elasticsearch.yml

Reviewer: The other lines for the snippet should be removed since they refer to the previous version of the `docker-compose.yml` file.

In order to avoid confusion if we'll update it again, what about adding something like this:

```
----
    volumes:
      - ./es01.yml:/usr/share/elasticsearch/config/elasticsearch.yml
      - ...
```

Same in the Kibana snippet.

Author: ++

are also available from the
https://github.com/elastic/stack-docs/blob/master/docs/en/getting-started/docker/[elastic/stack-docs]
repository on GitHub.

--
ifeval::["{release-state}"=="unreleased"]
NOTE: Version {version} of {es} has not been released,
so the sample Docker Compose and configuration files are not yet available for
this version. See the {stack-gs-current}/get-started-docker.html[current version]
for the latest sample files.
endif::[]
--

--
ifeval::["{release-state}"!="unreleased"]

[discrete]
[[docker-env-file]]
===== `.env`

The `.env` file sets environment variables that are used when you run the
`docker-compose.yml` configuration file. Ensure that you specify a strong
password for the `elastic` and `kibana_system` users with the
`ELASTIC_PASSWORD` and `KIBANA_PASSWORD` variables. These variable are
referenced by the `docker-compose.yml` file.

["source","txt",subs="attributes"]
----
include::docker/.env[]
----

[discrete]
[[docker-compose-file]]
===== `docker-compose.yml`

This `docker-compose.yml` file creates a three-node secure {es} cluster with authentication and network encryption enabled, and a {kib} instance securely connected to it.

> .Exposing ports on Linux

Reviewer: Are we sure this is linux-specific only?

Author: In the original issue, @pugnascotia helped to [write this initial description](https://github.com/elastic/elasticsearch/pull/48142#discussion_r357124521). It seems that this issue is specific to Linux and how Docker bypasses the UFW firewall by editing the `iptables` configuration directly.

Reviewer: If this is the standard behavior for docker even on other platforms (e.g. Windows), I'd rather be generic and mention the risk. We cannot assume that firewalls are enabled on hosts, so better be cautios.

Reviewer: LGTM if we replace `bound to the host` with `bound to localhost`.

Author: Sounds good to me -- included that change in 4a16b841ccb0fbbc51c2f49d6e612dfd5b38cbff

****
This configuration exposes port `9200` on all network interfaces. Given how
Docker manipulates `iptables` on Linux, this means that your {es} cluster is
publicly accessible, potentially ignoring any firewall settings. If you don't

> want to expose port `9200` and instead use a reverse proxy, replace `9200:9200`
> with `localhost:9200:9200` in the `docker-compose.yml` file. {es} will then only

Reviewer: ```suggestion
want to expose port `9200` to external hosts, set the ES_PORT value to something
like `127.0.0.1:9200` in the `.env` file. {es} will then only
```

Author: I'll make this change locally and include it in my next commit:

>If you don't want to expose port `9200` to external hosts, set the value for `ES_PORT` in 
the `.env` file to something like `127.0.0.1:9200`.

be accessible from the host machine itself.
****

[source,yaml,subs="attributes"]
----
include::docker/docker-compose.yml[]
----

endif::[]
--

[discrete]
==== Start your cluster with security enabled and configured

. Modify the `.env` file and enter strong password values for both the
`ELASTIC_PASSWORD` and `KIBANA_PASSWORD` variables.
+
NOTE: You must use the `ELASTIC_PASSWORD` value for further interactions with
the cluster. The `KIBANA_PASSWORD` value is only used internally when
configuring {kib}.

. Create and start the three-node {es} cluster and {kib} instance:
+
["source","sh"]
----
docker-compose up -d
----

. When the deployment has started, open a browser and navigate to http://localhost:5601[http://localhost:5601] to
access {kib}, where you can load sample data and interact with your cluster.

[[docker-compose-remove-containers]]
[discrete]

> ==== Remove Docker Compose containers

Reviewer: ```suggestion
==== Shut down and clean up the deployment
```

Author: I'll change this to **Stop and remove the deployment**.

>       - ./es-node01.yml:/usr/share/elasticsearch/config/elasticsearch.yml

Reviewer: ```suggestion
      - ./es01.yml:/usr/share/elasticsearch/config/elasticsearch.yml
```

We should also remove the previous three lines.

----

> Similarly, to load {kib} settings from a file, you overwrite
> `/usr/share/kibana/config/kibana.yml`:

Reviewer: ```suggestion
Similarly, to load {kib} settings from a file, you can add the following mount in the **volumes** section for the **kibana** service.
```

["source","yaml"]

> ----

Reviewer: ```suggestion
----
      - ./kibana.yml:/usr/share/kibana/config/kibana.yml
```

We also need to remove the following three lines.
